src/ensembl/src/content/app/entity-viewer/gene-view/components/protein-domain-image/ProteinDomainImage.tsx

```diff
@@ -0,0 +1,178 @@
+import React from 'react';
+import classNames from 'classnames';
+import { scaleLinear, ScaleLinear } from 'd3';
+
+import { getFeatureCoordinates } from 'src/content/app/entity-viewer/shared/helpers/entity-helpers';
+import { Transcript } from 'src/content/app/entity-viewer/types/transcript';
+import { ProteinDomainsResources } from 'src/content/app/entity-viewer/types/product';
+
+import styles from './ProteinDomainImage.scss';
+
+const BLOCK_HEIGHT = 18;
+const BACKBONE_HEIGHT = 24;
+
+export type ProteinDomainImageProps = {
+  transcript: Transcript;
+  width: number; // available width for drawing, in pixels
+  classNames?: {
+    backbone?: string;
+    exon?: string;
+    domain: string;
+  };
+};
+
+type ProteinDomainImageData = {
+  [resource_type: string]: {
+    [resource_description: string]: {
+      start: number;
+      end: number;
+    }[];
+  };
+};
+
+export const getDomainsByResourceGroups = (
+  proteinDomainsResources: ProteinDomainsResources
+) => {
+  const proteinDomains: ProteinDomainImageData = {};
+
+  Object.values(proteinDomainsResources).map((resource) => {
+    const resourceName = resource.name;
+
+    if (!proteinDomains[resourceName]) {
+      proteinDomains[resourceName] = {};
+    }
+
+    proteinDomainsResources[resourceName].domains.map((domain) => {
+      const domainName = domain.name;
+
+      const { start, end } = domain.location;
+
+      if (!proteinDomains[resourceName][domainName]) {
+        proteinDomains[resourceName][domainName] = [];
+      }
+
+      proteinDomains[resourceName][domainName].push({
+        start: start,
+        end: end
+      });
+    });
+  });
+
+  return proteinDomains;
+};
+
+const ProteinDomainImage = (props: ProteinDomainImageProps) => {
+  const { start: transcriptStart } = getFeatureCoordinates(props.transcript);
+
+  if (!props.transcript.product.protein_domains_resources) {
+    return null;
+  }
+
+  const length = props.transcript.product?.length || 0;
```

Comment:
Transcript type says that product is mandatory? Why the optional chaining then?

Reply:
Removed the optional chaining